src/persist-client/src/write.rs

```diff
@@ -403,57 +685,82 @@ where
         );
 
         let batch = BlobTraceBatchPart {
-            desc,
-            updates: builder.finish(),
+            desc: encoded_desc.clone(),
+            updates,
             index: 0,
         };
-        if batch.updates.len() == 0 {
-            return Ok(None);
-        }
 
         let mut buf = Vec::new();
         batch.encode(&mut buf);
-        Ok(Some(buf))
-    }
 
-    /// Test helper for an [Self::append] call that is expected to succeed.
-    #[cfg(test)]
-    #[track_caller]
-    pub async fn expect_append<L, U>(&mut self, updates: &[((K, V), T, D)], lower: L, new_upper: U)
-    where
-        L: Into<Antichain<T>>,
-        U: Into<Antichain<T>>,
-    {
-        self.append(
-            updates.iter().map(|((k, v), t, d)| ((k, v), t, d)),
-            lower.into(),
-            new_upper.into(),
-        )
-        .await
-        .expect("external durability failed")
-        .expect("invalid usage")
-        .expect("unexpected upper");
+        let key = Uuid::new_v4().to_string();
+        let () = retry_external("compare_and_append::set", || async {
+            // If MultiBlob::set took value as a ref, then we wouldn't have
+            // to clone here.
+            self.blob
+                .set(
+                    Instant::now() + FOREVER,
+                    &key,
+                    buf.clone(),
+                    Atomicity::RequireAtomic,
+                )
+                .await
+        })
+        .await;
+
+        let batch = Batch::new(
+            Arc::clone(&self.blob),
+            self.shard_id.clone(),
+            desc,
+            vec![key],
+        );
+
+        Ok(batch)
     }
 
-    /// Test helper for a [Self::compare_and_append] call that is expected to
-    /// succeed.
-    #[cfg(test)]
-    #[track_caller]
-    pub async fn expect_compare_and_append(
-        &mut self,
-        updates: &[((K, V), T, D)],
-        expected_upper: T,
-        new_upper: T,
-    ) {
-        self.compare_and_append(
-            updates.iter().map(|((k, v), t, d)| ((k, v), t, d)),
-            Antichain::from_elem(expected_upper),
-            Antichain::from_elem(new_upper),
-        )
-        .await
-        .expect("external durability failed")
-        .expect("invalid usage")
-        .expect("unexpected upper")
+    /// Adds the given update to the batch.
+    ///
+    /// The update timestamp must be greater or equal to `lower` that was given
+    /// when creating this [BatchBuilder].
+
```

Comment:
oh this might be my fault. I think we still need some sort of `//` or `///` prefix here before the empty line for rustdoc to work :)